File: adaria-welcome/src/aci.js
// Conector ACI Dali (SQL Server) - SOLO LECTURA.
// Usuario adaria_ro tiene permisos SELECT unicamente sobre AdariaVeraHotel.
// Este modulo NUNCA debe ejecutar INSERT/UPDATE/DELETE.
const sql = require('mssql');

const config = {
  server: process.env.ACI_HOST,
  port: parseInt(process.env.ACI_PORT || '1433', 10),
  database: process.env.ACI_DATABASE,
  user: process.env.ACI_USER,
  password: process.env.ACI_PASSWORD,
  options: {
    encrypt: false,
    trustServerCertificate: true
  },
  connectionTimeout: 15000,
  requestTimeout: 15000,
  pool: { max: 5, min: 0, idleTimeoutMillis: 30000 }
};

let poolPromise = null;

function getPool() {
  if (!poolPromise) {
    poolPromise = new sql.ConnectionPool(config).connect();
  }
  return poolPromise;
}

// Llegadas de hoy: Reservas (res_ent_dat = hoy, res_anu_bln = 0)
// JOIN Habitaciones (HAB_GUID), ReservaHuespedes/Huespedes (HUE_GUID), Naciones (NAC_GUID)
// Agrupado por reserva con su lista de huespedes.
const ARRIVALS_QUERY = `
  SELECT
    r.RES_GUID          AS resGuid,
    r.RES_COD_str       AS resCod,
    r.res_ent_dat        AS checkinDate,
    r.res_sal_dat        AS checkoutDate,
    ISNULL(h.HAB_COD_str, NULL) AS roomCode,
    hu.HUE_GUID          AS hueGuid,
    hu.hue_des_str       AS nombre,
    hu.hue_co1_str       AS apellido1,
    hu.hue_co2_str       AS apellido2,
    n.nac_des_str        AS nacionalidad
  FROM Reservas r
  LEFT JOIN Habitaciones h      ON h.HAB_GUID = r.HAB_GUID
  LEFT JOIN ReservaHuespedes rh ON rh.RES_GUID = r.RES_GUID
  LEFT JOIN Huespedes hu        ON hu.HUE_GUID = rh.HUE_GUID
  LEFT JOIN Naciones n          ON n.NAC_GUID = hu.NAC_GUID
  WHERE CAST(r.res_ent_dat AS DATE) = CAST(GETDATE() AS DATE)
    AND r.res_anu_bln = 0
  ORDER BY r.RES_GUID, hu.HUE_GUID
`;

async function getArrivalsToday() {
  const pool = await getPool();
  const result = await pool.request().query(ARRIVALS_QUERY);

  const reservasMap = new Map();
  for (const row of result.recordset) {
    if (!reservasMap.has(row.resGuid)) {
      reservasMap.set(row.resGuid, {
        resGuid: row.resGuid,
        resCod: (row.resCod || '').trim(),
        checkinDate: row.checkinDate,
        checkoutDate: row.checkoutDate,
        roomCode: row.roomCode ? row.roomCode.trim() : null,
        huespedes: []
      });
    }
    if (row.hueGuid) {
      reservasMap.get(row.resGuid).huespedes.push({
        hueGuid: row.hueGuid,
        nombre: (row.nombre || '').trim(),
        apellido1: (row.apellido1 || '').trim(),
        apellido2: (row.apellido2 || '').trim(),
        nacionalidad: (row.nacionalidad || '').trim()
      });
    }
  }
  return Array.from(reservasMap.values());
}

async function testConnection() {
  const pool = await getPool();
  await pool.request().query('SELECT 1 AS ok');
  return true;
}

module.exports = { getArrivalsToday, testConnection };

File: adaria-ine/aci.py
# -*- coding: utf-8 -*-
"""
Acceso de SOLO LECTURA al PMS ACI Dali (SQL Server / ACIGRUP) para el
Hotel Adaria Vera, usado por el módulo INE (encuesta EOH).

NUNCA se ejecuta ninguna escritura contra ACI desde este módulo. Todas las
consultas son SELECT sobre las tablas: Reservas, ReservaHuespedes,
Huespedes, Naciones y Habitaciones.
"""
import os
import calendar
from datetime import date

import pymssql

ACI_HOST = os.environ.get("ACI_HOST", "192.168.1.34")
ACI_PORT = int(os.environ.get("ACI_PORT", "1433"))
ACI_USER = os.environ.get("ACI_USER", "adaria_ro")
ACI_PASSWORD = os.environ.get("ACI_PASSWORD", "")
ACI_DATABASE = os.environ.get("ACI_DATABASE", "AdariaVeraHotel")


def get_connection():
    """Abre una conexión de solo lectura a ACI. El usuario adaria_ro no
    tiene permisos de escritura en el PMS, pero además esta capa nunca
    ejecuta INSERT/UPDATE/DELETE."""
    return pymssql.connect(
        server=ACI_HOST,
        port=ACI_PORT,
        user=ACI_USER,
        password=ACI_PASSWORD,
        database=ACI_DATABASE,
        login_timeout=10,
        timeout=30,
        as_dict=True,
    )


def month_bounds(anio: int, mes: int):
    """Devuelve (primer_dia, primer_dia_mes_siguiente) como date."""
    primer_dia = date(anio, mes, 1)
    ultimo_dia_num = calendar.monthrange(anio, mes)[1]
    if mes == 12:
        siguiente = date(anio + 1, 1, 1)
    else:
        siguiente = date(anio, mes + 1, 1)
    return primer_dia, siguiente, ultimo_dia_num


def get_habitaciones_disponibles():
    """Nº de habitaciones no obsoletas dadas de alta en ACI (referencia
    para el campo manual 'habitaciones disponibles')."""
    conn = get_connection()
    try:
        cur = conn.cursor()
        cur.execute(
            "SELECT COUNT(*) AS n FROM Habitaciones "
            "WHERE HAB_OBSOLETA = 0 OR HAB_OBSOLETA IS NULL"
        )
        row = cur.fetchone()
        return int(row["n"]) if row else 0
    finally:
        conn.close()


# Pernoctaciones = suma de noches de cada huésped dentro del mes seleccionado
# (solape de [res_ent_dat, res_sal_dat) con [primer_dia, primer_dia_mes_siguiente)).
# Viajeros = huéspedes cuya fecha de ENTRADA cae dentro del mes seleccionado.
#
# ReservaHuespedes no está sistemáticamente poblada: para reservas sin fila
# en ReservaHuespedes se usa como único huésped el HUE_GUID titular de
# Reservas (comprobado: cuando ReservaHuespedes SÍ tiene filas, el titular
# ya está incluido entre ellas, así que no hay doble conteo).
_QUERY_NACIONALIDADES = """
WITH Guests AS (
    SELECT rh.RES_GUID, rh.HUE_GUID FROM ReservaHuespedes rh
    UNION
    SELECT r.RES_GUID, r.HUE_GUID FROM Reservas r
    WHERE NOT EXISTS (SELECT 1 FROM ReservaHuespedes rh2 WHERE rh2.RES_GUID = r.RES_GUID)
)
SELECT
    ISNULL(n.nac_des_str, 'SIN DETERMINAR') AS pais,
    n.nac_abr_str AS codigo,
    COUNT(CASE WHEN r.res_ent_dat >= %(ms)s AND r.res_ent_dat < %(me)s THEN 1 END) AS viajeros,
    SUM(CASE WHEN r.res_sal_dat > %(ms)s AND r.res_ent_dat < %(me)s
        THEN DATEDIFF(day,
               CASE WHEN r.res_ent_dat > %(ms)s THEN r.res_ent_dat ELSE %(ms)s END,
               CASE WHEN r.res_sal_dat < %(me)s THEN r.res_sal_dat ELSE %(me)s END)
        ELSE 0 END) AS pernoctaciones
FROM Guests g
JOIN Reservas r ON r.RES_GUID = g.RES_GUID
JOIN Huespedes h ON h.HUE_GUID = g.HUE_GUID
LEFT JOIN Naciones n ON n.NAC_GUID = h.NAC_GUID
WHERE r.res_anu_bln = 0
GROUP BY n.nac_des_str, n.nac_abr_str
HAVING COUNT(CASE WHEN r.res_ent_dat >= %(ms)s AND r.res_ent_dat < %(me)s THEN 1 END) > 0
    OR SUM(CASE WHEN r.res_sal_dat > %(ms)s AND r.res_ent_dat < %(me)s
        THEN DATEDIFF(day,
               CASE WHEN r.res_ent_dat > %(ms)s THEN r.res_ent_dat ELSE %(ms)s END,
               CASE WHEN r.res_sal_dat < %(me)s THEN r.res_sal_dat ELSE %(me)s END)
        ELSE 0 END) > 0
ORDER BY pernoctaciones DESC, viajeros DESC
"""


def get_pernoctaciones_por_nacionalidad(anio: int, mes: int):
    """Devuelve lista de dicts {pais, codigo, viajeros, pernoctaciones}
    para el mes indicado, y los totales agregados."""
    primer_dia, siguiente, _ = month_bounds(anio, mes)
    conn = get_connection()
    try:
        cur = conn.cursor()
        cur.execute(_QUERY_NACIONALIDADES, {"ms": primer_dia, "me": siguiente})
        filas = cur.fetchall()
    finally:
        conn.close()

    resultado = []
    total_viajeros = 0
    total_pernoctaciones = 0
    for f in filas:
        viajeros = int(f["viajeros"] or 0)
        pernoctaciones = int(f["pernoctaciones"] or 0)
        total_viajeros += viajeros
        total_pernoctaciones += pernoctaciones
        resultado.append({
            "pais": f["pais"],
            "codigo": (f["codigo"] or "").strip(),
            "viajeros": viajeros,
            "pernoctaciones": pernoctaciones,
        })
    return {
        "filas": resultado,
        "total_viajeros": total_viajeros,
        "total_pernoctaciones": total_pernoctaciones,
    }
